README.md
# `apvi-webservice`

Accesses data from APVI to produce CSVs and charts of solar energy production for use on AREMI.

## Building
1. Follow the instructions on http://docs.haskellstack.org/en/stable/README.html on how to install `stack`
2. From this directory, run `stack setup` if you havcen't run `stack` before, or if it asks you to.
3. Run `stack build` to build the above three libraries.



Changelog.md
### apvi-webservice version 0.10.0.0
- Update to new configuration format. See config/apvi-webservice.yaml and `--help` in the executable

### apvi-webservice version 0.9.1.1
- Fix RTS options in cabal file.

### apvi-webservice version 0.9.0.0
- Make a best effort attempt to make new charts if possible - if we can't make new charts, the old ones will remain available.

### aremi-web-utils version 0.9.0.0
- Added a check for charts which have a single x/y pair - when x is a `LocalTime` this causes unbounded memory usage until the app crashes. Now we just fail and keep trying until we can actually make a graph.

## Package split

Packages for `apvi-webservice`, `aemo-webservice` and `aremi-web-utils` have been split out into their own packages. Changelog will remain in this file from this point on, but changes will be in the form `### <component> version a.b.c.d`

Building is the same as before, run `stack build` from this directory.

### Version 0.8.2.1
- Change "Most Recent" to "Current" for generation values (leaving "Most Recent Sample Time" as is).

### Version 0.8.2.0
- Use APVI API key.

### Version 0.8.1.0
- Remove APVI initialisation so service can launch when APVI are down, returning 404s for requests

### Version 0.8.0.0
 - Adds ability to download information about a particular DUID as a CSV using either `{start,end}Time=<date as "%Y-%M-%DT%h:%m:%s%z">`, eg. "2015-08-08T01:00:00+1000" or using `offset=1Y2M3D4h5m`

  Other:
   - Add links to download DUID data as CSV, add "renewables" category available at `/aemo/v4/csv/renewables`
   - Export and add ToText instance for TimePeriod to make queries using safeLink
   - Add support for specifying an offset to get DUID CSV data back to
   - Allow DUID CSV data to specify start and end times as URL params
   - Add ISOUtcTime to parse ISO8601 formatted timestamps in URLs
   - Bump version to 0.8.0.0, fix call to --with-rtsopts in ghc-options
   - Add v4 API which allows downloading of CSV of all data for every DUID
   - Reduce frequency of idle time full GC to reduce CPU usage
   - Remove comment and redundant let for duid in URL
   - Don't URL encode things (happens automatically), add deps to 2.19 stack.yaml

  Contributors:
   - Alex Mason

### Version 0.7.1.3
- Rename Time column since it interacts badly with Terria.
- Improve other column names.

### Version 0.7.1.2
- Update make_package to include the new fonts path.

### Version 0.7.1.1
- Forgotten liftIO import fixed

### Version 0.7.1.0
- Font location can now be configured in the config using "font-path" key, defaults to ./fonts
- Fonts are only loaded once at app launch

### Version 0.7.0.3
- New optional API V3 config file (for running V2 and V3 concurrently on different ports).
- Make make_package use Stack.

  Other:
   - Also look for a v3 config file in order to run concurrently with API v2.
   - Change make_package to use Stack instead of Cabal.
   - Only warn when making a package and there are local modifications.
   - Fixing CPP macro indentation for Linux build.
   - Point to LTS-2.19.
   - Fix maccidental removal of lts-3.0 yaml file

### Version 0.7.0.0
- v3 of AEMO live generation API now supports creating PNGs instead of SVGs to reduce bandwidth and rendering time. APVI services will be converted to use PNGs in the next major version.
- Use `stack` for building
- Now compatible with ghc 7.10 - change the symlink for stack.yaml between the two stack.lts-{2.19,3.0}.yaml files
- Add Gzip support because why not

  Other:
  - Add stack-lts-{2.19,3.0}.yaml and make stack.yaml a symlink to 2.19
   - Add "other-modules:", fix library versions for lts-3.0
   - Add gzip support to wai middleware
   - Replace SVGs with PNGs, ghc 7.10, http-conduit 2.1.6, chart-diagrams 1.3 and time 1.5 support
   - Add DEnv part of the environment for LivePower so it's not created every time we need an image, make compatible with chart-diagrams 1.2 and 1.3, and ghc 7.10
   - Make Util.Charts compatible with 7.10
   - Add servePNG and make Util.Web compatible with 7.10
   - Remove tabs, make Util.Fetch compatible with 7.10
   - Remove tabs and add impossible [] case in Util.Periodic
   - Move MimeRender instance for Tagged k ByteString into Util.Types
   - Make use of Data.Tagged in `servant` and replace newtypes with `Tagged CSV ByteString` etc.
   - Ignore .stack-work/, add comments to functions
   - Remove import of renderableToSVGString in Util.Charts - Thanks --pedantic
   - Bump version, update changelog, add missing PNG types to Util.Types
   - Add JuicyPixels dep
   - Pull out renderImage into Util.Charts
   - Implement initial rasterific rendering and update API versions
   - Make renderPSDChart pure - it wasn't using IO
   - Convert to using `stack`
   - Use 'git status --porcelain' instead of complicated diff to check for local modifications.
   - Ensure make_package doesn't run on anything but a Linux host.
   - Update version number and changelog.
   - Added percentage of Max Cap calculated value.
   - Change graph dataseries lines from 3 to 2 "pixel" thickness.

  Contributors:
   - Mats Henrikson
   - Alex Mason

### Version: 0.6.2.1
- Added percentage of Max Cap calculated value.
- Change graph dataseries lines from 3 to 2 "pixel" thickness.
- Chart-diagrams package version changed.

### Version: 0.6.2.0
- Add a column to the AEMO CSV output which shows the percentage of current generation compared to the reg cap

### Version: 0.6.1.0
Changes:
- `AEMO-archiver` library became `aemo-archiver`

### Version 0.6.0.0
Major changes:
- API now includes version numbers

Other (auto):
   - Adding 'add-source' to show dependency on aemo-archiver. Making the package artifact use a better name. Don't make a package if there are uncommitted changes.
   - Minor version increment to deal with new aemo-archiver lib name.
   - Ignoring the target directory.

  Contributors:
   - Mats Henrikson
   - Alex Mason


### Version 0.5.0.0

Major changes:

src/Main.hs
{-# LANGUAGE BangPatterns          #-}
{-# LANGUAGE DataKinds             #-}
{-# LANGUAGE FlexibleInstances     #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE OverloadedStrings     #-}
{-# LANGUAGE TemplateHaskell       #-}
{-# LANGUAGE TypeOperators         #-}
{-# LANGUAGE TypeSynonymInstances  #-}
{-# LANGUAGE GeneralizedNewtypeDeriving #-}

module Main where

import           System.Log.Formatter                      (simpleLogFormatter)
import           System.Log.Handler                        (setFormatter)
import           System.Log.Handler.Simple
import qualified System.Log.Logger                         as HSL
import           System.Log.Logger.TH                      (deriveLoggers)

import           Network.Wai                               (Middleware)
import           Network.Wai.Handler.Warp                  (run)
import           Network.Wai.Middleware.Cors               (simpleCors)
import           Network.Wai.Middleware.Gzip               (gzip)
import           Network.Wai.Middleware.RequestLogger      (Destination (..),
                                                            IPAddrSource (..),
                                                            OutputFormat (..), RequestLoggerSettings (..),
                                                            mkRequestLogger)
import           Network.Wai.Util                          (replaceHeader)
import           Servant
import           Servant.HTML.Blaze

import           System.IO                                 (BufferMode (..),
                                                            IOMode (..),
                                                            hSetBuffering,
                                                            openFile)

import           Control.Monad.IO.Class                    (liftIO)
import           Control.Monad.Trans.Either

import           Data.Default

import           APVI.LiveSolar

import           Graphics.Rendering.Chart.Backend.Diagrams (createEnv)
import           Graphics.Rendering.Chart.Easy             hiding (Default)
import           Util.Charts                               (loadFonts)
import           Util.Types

import           Data.Text.Encoding                        (encodeUtf8)

import           Network.Wai.Metrics
import qualified System.Remote.Monitoring                  as M

import           Configuration.Utils
import           PkgInfo_apvi_webservice


$(deriveLoggers "HSL" [HSL.DEBUG, HSL.INFO, HSL.ERROR, HSL.WARNING])

type App =
         "apvi" :> (APVILiveSolar :<|> "api" :> Get '[HTML] APIDoc)
    :<|> "static" :> Raw


appProxy :: Proxy App
appProxy = Proxy

appServer :: APVIConf -> EitherT String IO (Server App)
appServer conf = do
    fontSelector <- liftIO $ loadFonts (_apviFontPath conf)
    let !env = createEnv bitmapAlignmentFns 500 300 fontSelector
    ls <- EitherT $ makeLiveSolarServer conf env
    return $ (ls :<|> return (docsHtml "APVI WebService API" appProxy))
            :<|> serveDirectory "static"
    where
        _addCorsHeader :: Middleware
        _addCorsHeader app req respond = app req (respond . replaceHeader ("Access-Control-Allow-Origin","*"))


makeMiddleware :: APVIConf -> IO Middleware
makeMiddleware config = do
    let accessLog = _apviAccessLog config
        mmonPort  = _apviMonitoringPort config
        mmonHost  = encodeUtf8 <$> _apviMonitoringHost config
    h <- openFile accessLog AppendMode
    hSetBuffering h NoBuffering
    accessLogger <- mkRequestLogger (def {destination = Handle h
                                         ,outputFormat = Apache FromFallback
                                         ,autoFlush = True})
    monitor <- case (,) <$> mmonPort <*> mmonHost of
        Nothing -> pure id
        Just (monPort,monHost) -> do
            store <- M.serverMetricStore <$> M.forkServer monHost monPort
            waiMetrics <- registerWaiMetrics store
            pure $ metrics waiMetrics
    return (monitor . accessLogger . gzip def . simpleCors)
    -- return (logStdoutDev . simpleCors)
    -- return (simpleCors)


mainInfo :: ProgramInfo APVIConf
mainInfo = programInfo "APVI Webservice" pAPVIConf defaultApviConf

main :: IO ()
main = runWithPkgInfoConfiguration mainInfo pkgInfo $ \config -> do
    let appLog = _apviAppLog config
    h' <- fileHandler appLog HSL.DEBUG
    h <- return $ setFormatter h' (simpleLogFormatter "[$time] $prio $loggername: $msg")
    HSL.updateGlobalLogger "Main" (HSL.addHandler h . HSL.setLevel HSL.DEBUG)
    HSL.updateGlobalLogger "APVI.LiveSolar" (HSL.addHandler h . HSL.setLevel HSL.DEBUG)
    infoM "apvi-webservice launch"

    appServ <- runEitherT (appServer config)

    case appServ of
        Left err -> errorM err
        Right serv -> do
            let port = _apviHttpPort config
            mids <- makeMiddleware config
            run port $ mids $ serve appProxy serv


src/APVI/Types.hs
{-# LANGUAGE DataKinds                  #-}
{-# LANGUAGE FlexibleInstances          #-}
{-# LANGUAGE GeneralizedNewtypeDeriving #-}
{-# LANGUAGE MultiParamTypeClasses      #-}
{-# LANGUAGE OverloadedStrings          #-}
{-# LANGUAGE StandaloneDeriving         #-}
{-# LANGUAGE TemplateHaskell            #-}

module APVI.Types where

import           Servant
import           Servant.Docs


import           Data.Aeson

import           Data.HashMap.Strict  (HashMap)
import qualified Data.HashMap.Strict  as H
import           Data.List            (intercalate)
import           Data.Text            (Text)
import qualified Data.Text            as T

import           Network.HTTP.Conduit (Manager)

import           Control.Lens

import           Data.Default         (Default (..))

import qualified Data.Vector          as V


import           APVI.Docs

import           Util.Charts
import           Util.Types


states :: [(Text, Int)]
states = [
    ("nsw",1),
    ("vic",2),
    ("qld",3),
    ("sa",4),
    ("wa",5),
    ("tas",6)
    -- ("nt",7)
    ]



data AppState = AppState {
    _latestETag              :: !(Maybe ETag)
    , _contributionCSV       :: !(Maybe (Text -> CsvBS))
    , _contributionGraphs    :: !(HashMap Text PngBS)
    , _performanceCSV        :: !(Maybe (Text -> CsvBS))
    , _performanceGraphs     :: !(HashMap Text PngBS)
    , _performanceGraphJSON  :: !Value
    , _contributionGraphJSON :: !Value
    , _chartEnv              :: !(Maybe ChartEnv)
    , _httpManager           :: !(Maybe Manager)
}

$(makeLenses ''AppState)


instance Default AppState where
    def = AppState {
        _latestETag            = Nothing,
        _contributionCSV       = Nothing,
        _contributionGraphs    = H.empty,
        _performanceCSV        = Nothing,
        _performanceGraphs     = H.empty,
        _performanceGraphJSON  = Array V.empty,
        _contributionGraphJSON = Array V.empty,
        _chartEnv              = Nothing,
        _httpManager           = Nothing
    }



newtype PerformancePNG   = PPNG  {unPPNG   :: PngBS} deriving (Eq, Show, MimeRender PNG)
newtype PerformanceCSV   = PCSV  {unPCSV   :: CsvBS} deriving (Eq, Show, MimeRender CSV)
newtype PerformanceJSON  = PJSON {unPJSON  :: Value} deriving (Eq, Show)
newtype ContributionPNG  = CPNG  {unCPNG   :: PngBS} deriving (Eq, Show, MimeRender PNG)
newtype ContributionCSV  = CCSV  {unCCSV   :: CsvBS} deriving (Eq, Show, MimeRender CSV)
newtype ContributionJSON = CJSON {unCJSON  :: Value} deriving (Eq, Show)

newtype StateName        = SN {unStateName :: Text}  deriving (Eq, Show)

instance ToSample PerformancePNG PerformancePNG where toSample _ = Just (PPNG $ Tagged "(A PNG Image)")
instance ToSample PerformanceCSV PerformanceCSV where toSample _ = Just . PCSV . Tagged $ perfCsvSample
instance ToSample PerformanceJSON Text where toSample _ = Just $ perfJsonSample
instance ToSample ContributionPNG ContributionPNG where toSample _ = Just (CPNG $ Tagged "(A PNG Image)")
instance ToSample ContributionCSV ContributionCSV where toSample _ = Just . CCSV . Tagged $ contCsvSample
instance ToSample ContributionJSON Text where toSample _ = Just $ contJsonSample
deriving instance {-# OVERLAPPING #-} MimeRender JSON PerformanceJSON
deriving instance {-# OVERLAPPING #-} MimeRender JSON ContributionJSON
instance ToSample StateName Text where toSample _ = Just "nsw"
instance FromText StateName where fromText = Just . SN

instance ToCapture (Capture "pngstate" StateName) where
    toCapture _ = DocCapture "State name"
                    $ "The name of the state to produce a chart for. Values may be 'all' or one of: "
                    ++ intercalate ", " (map (T.unpack . fst) states)
                    ++ "."



src/APVI/LiveSolar.hs
{-# LANGUAGE CPP                        #-}
{-# LANGUAGE BangPatterns               #-}
{-# LANGUAGE DataKinds                  #-}
{-# LANGUAGE DeriveGeneric              #-}
{-# LANGUAGE FlexibleInstances          #-}
{-# LANGUAGE GeneralizedNewtypeDeriving #-}
{-# LANGUAGE MultiParamTypeClasses      #-}
{-# LANGUAGE OverloadedStrings          #-}
{-# LANGUAGE PolyKinds                  #-}
{-# LANGUAGE QuasiQuotes                #-}
{-# LANGUAGE RankNTypes                 #-}
{-# LANGUAGE StandaloneDeriving         #-}
{-# LANGUAGE TemplateHaskell            #-}
{-# LANGUAGE TupleSections              #-}
{-# LANGUAGE TypeFamilies               #-}
{-# LANGUAGE TypeOperators              #-}


module APVI.LiveSolar (
    -- Configuration
    APVIConf(..),
    pAPVIConf,
    defaultApviConf,
    apviFontPath,
    apviAccessLog,
    apviAppLog,
    apviHttpPort,
    apviMonitoringPort,
    apviMonitoringHost,
    apviUpdateFreqSecs,
    apviRetries,
    -- Web interface
    APVILiveSolar,
    makeLiveSolarServer,
    --Types
    PngBS,
    CsvBS,
    AppState(..),
    --
    updateRef,
    every,
    isErr,
    initialiseLiveSolar,
    -- Lenses
    contributionGraphs,
    performanceGraphs,
    contributionCSV,
    performanceCSV
    ) where


import           Data.List                     (sortBy)
import           Data.Maybe                    (catMaybes)
import           Data.Ord                      (comparing)

import           Control.Applicative
import           Control.Arrow                 (second)

import qualified Data.ByteString               as S
import qualified Data.ByteString.Lazy          as BSL

import           Data.Text                     (Text)
import qualified Data.Text                     as T

import           Data.Text.Encoding            (encodeUtf8)


import qualified Data.HashMap.Strict           as H

import           Data.Csv                      (defaultEncodeOptions,
                                                encodeByNameWith, toField)
import qualified Data.Csv                      as Csv
import qualified Data.Vector                   as V

import           Control.Lens                  as L
import qualified Data.Aeson                    as A
import           Data.Aeson.Lens               as AL
import           Data.Text.Lens

import           Graphics.Rendering.Chart.Easy hiding (Default, (.=))

import           Data.Time.Clock               (UTCTime)
#if MIN_VERSION_time(1,5,0)
import           Data.Time.Format              (defaultTimeLocale, formatTime,
                                                parseTimeM)
#else
import           Data.Time.Format              (formatTime, parseTime)
import           System.Locale                 (defaultTimeLocale)
#endif
import           Data.Time.LocalTime           (LocalTime (..), TimeZone,
                                                hoursToTimeZone, timeZoneName,
                                                utcToLocalTime)

import qualified System.Log.Logger             as HSL
import           System.Log.Logger.TH          (deriveLoggers)


import           Data.IORef                    (IORef, readIORef, writeIORef)

import           Control.Monad.Catch           (SomeException, catch)
import           Control.Monad.IO.Class

#if MIN_VERSION_http_conduit(2,1,6)
import           Network.HTTP.Conduit          (newManager, tlsManagerSettings)
#else
import           Network.HTTP.Conduit          (withManager)
#endif

import           Control.Concurrent.Async      (concurrently)
import           Control.Retry                 (fibonacciBackoff, limitRetries,
                                                retrying)
import           Data.IORef                    (newIORef)
import           Data.Time.Units               hiding (Day)

import           Servant

import           Util.Charts
import           Util.Fetch
import           Util.Periodic
import           Util.Types
import           Util.Web

import           Text.Printf                   (printf)

import           APVI.Types

import           Configuration.Utils

$(deriveLoggers "HSL" [HSL.DEBUG, HSL.ERROR, HSL.WARNING, HSL.INFO])

data APVIConf = APVIConf
  { _apviFontPath       :: FontConfig
  , _apviAccessLog      :: String
  , _apviAppLog         :: String
  , _apviHttpPort       :: Int
  , _apviMonitoringPort :: Maybe Int
  , _apviMonitoringHost :: Maybe Text
  , _apviUpdateFreqSecs :: Integer
  , _apviRetries        :: Int
  }
$(makeLenses ''APVIConf)

defaultApviConf :: APVIConf
defaultApviConf = APVIConf
  { _apviFontPath       = defaultFontConfig
  , _apviAccessLog      = "access.log"
  , _apviAppLog         = "apvi-webservice.log"
  , _apviHttpPort       = 3000
  , _apviMonitoringPort = Nothing
  , _apviMonitoringHost = Nothing
  , _apviUpdateFreqSecs = 5
  , _apviRetries        = 10
  }

instance FromJSON (APVIConf -> APVIConf) where
    parseJSON = withObject "APVIConf" $ \o -> id
      <$< apviFontPath       %.: "apviFontPath" % o
      <*< apviAccessLog      ..: "apviAccessLog" % o
      <*< apviAppLog         ..: "apviAppLog" % o
      <*< apviHttpPort       ..: "apviHttpPort" % o
      <*< apviMonitoringPort ..: "apviMonitoringPort" % o
      <*< apviMonitoringHost ..: "apviMonitoringHost" % o
      <*< apviUpdateFreqSecs ..: "apviUpdateFreqSecs" % o
      <*< apviRetries        ..: "apviRetries" % o

instance ToJSON APVIConf where
  toJSON a = object
    [ "apviFontPath"       A..= _apviFontPath a
    , "apviAccessLog"      A..= _apviAccessLog a
    , "apviAppLog"         A..= _apviAppLog a
    , "apviHttpPort"       A..= _apviHttpPort a
    , "apviMonitoringPort" A..= _apviMonitoringPort a
    , "apviMonitoringHost" A..= _apviMonitoringHost a
    , "apviUpdateFreqSecs" A..= _apviUpdateFreqSecs a
    , "apviRetries"        A..= _apviRetries a
    ]


pAPVIConf :: MParser APVIConf
pAPVIConf = id
  <$< apviFontPath       %:: pFontConfig
  <*< apviAccessLog      .:: strOption   % mkArg 'l' "access-log"  "LOGFILE"    "HTTP access log path"
  <*< apviAppLog         .:: strOption   % mkArg 'l' "app-log"     "APPLOGFILE" "HTTP app log path"
  <*< apviHttpPort       .:: option auto % mkArg 'p' "http-port"   "PORT"       "HTTP server port"
  <*< apviMonitoringPort .:: option auto % mkArg 'm' "mon-port"    "MONPORT"    "EKG monitoring HTTP port"
  <*< apviMonitoringHost .:: option auto % mkArg 'h' "mon-host"    "MONHOST"    "EKG monitoring HTTP hostname (eg. localhost, 127.0.0.1)"
  <*< apviUpdateFreqSecs .:: option auto % mkArg 'u' "update-freq" "MINS"       "How often to poll APVI for new data in minutes"
  <*< apviRetries        .:: option auto % mkArg 'r' "retries"     "RETRIES"    "How often to retry fetching data before giving up"
  where mkArg s l m h = short s <> long l <> metavar m <> help h

type APVILiveSolar = "v3" :> (
    "performance" :>
        (    "csv"                                  :> Header "Host" Text :> Get '[CSV] PerformanceCSV
        :<|> "png" :> Capture "pngstate" StateName  :> Get '[PNG] PerformancePNG
        :<|> "json"                                 :> Get '[JSON] PerformanceJSON)
    :<|>
    "contribution" :>
        (    "csv"                                  :> Header "Host" Text :> Get '[CSV] ContributionCSV
        :<|> "png" :> Capture "pngstate" StateName  :> Get '[PNG] ContributionPNG
        :<|> "json"                                 :> Get '[JSON] ContributionJSON)
    )


makeLiveSolarServer :: APVIConf -> ChartEnv ->  IO (Either String (Server APVILiveSolar))
makeLiveSolarServer conf env = do
    eref <- initialiseLiveSolar conf env
    case eref of
        Left s -> return $ Left s
        Right ref -> return $ Right (
            (        (fmap PCSV . serveCSV ref performanceCSV)
                :<|> (fmap PPNG . servePNG ref performanceGraphs . unStateName)
                :<|> (PJSON <$> serveJSON ref performanceGraphJSON))
            :<|> (   (fmap CCSV . serveCSV ref contributionCSV)
                :<|> (fmap CPNG . servePNG ref contributionGraphs . unStateName)
                :<|> (CJSON <$> serveJSON ref contributionGraphJSON))
            )


initialiseLiveSolar :: APVIConf -> ChartEnv -> IO (Either String (IORef AppState))
initialiseLiveSolar conf env = do
    manager <- newManager tlsManagerSettings
    ref <- newIORef def { _chartEnv = Just env, _httpManager = Just manager}
    let mins    = _apviUpdateFreqSecs conf
        retries = _apviRetries conf
    _tid <- updateRef retries ref `every` (fromInteger mins :: Minute)
    return $ Right ref
        -- else return $ Left "Failed to initialise live solar data"

-- type APVIPerformanceSVGPath = "aemo" :> "v2" :> "performance" :> "svg" :> Capture "state" Text
-- type APVIContributionSVGPath = "aemo" :> "v2" :> "contribution" :> "svg" :> Capture "state" Text

-- Takes a number of retries and the current app state ref and attempts to contact APVI for the latest
-- data for today.
updateRef :: Int -> IORef AppState -> IO Bool
updateRef retries ref = flip catch (\e -> (warningM  . show $ (e :: SomeException)) >> return False) $ do
    infoM "Starting update"
    let secretToken = "m181rsunPwre71io" :: String
        -- url = formatTime defaultTimeLocale "http://pv-map.apvi.org.au/data/%F" day
        url = printf "http://pv-map.apvi.org.au/api/v1/data/today.json?access_token=%s" secretToken
    current <- readIORef ref
    ejsn <- retrying (fibonacciBackoff 1000 <> limitRetries retries)
                     (\_ e -> return (isErr e))
                     $ \_ -> fetchFromCache (current ^?! httpManager . _Just) url (current ^. latestETag)
                        [("Accept"          , "application/json, text/javascript, */*"),
                         ("X-Requested-With", "XMLHttpRequest")
                        ]
    case ejsn of
        Err err -> errorM err >> return False
        NoChange -> debugM "update not necessary" >> return True
        NewData metag fetched bs -> updateRef' ref metag fetched bs


updateRef' :: IORef AppState -> Maybe ETag -> UTCTime -> BSL.ByteString -> IO Bool
updateRef' ref metag fetched bs = case A.eitherDecode' bs of
    Left err -> errorM err >> return False
    Right jsn -> do
        current <- readIORef ref
        let Just env = current ^. chartEnv
            tz = (hoursToTimeZone 10) {timeZoneName = "AEST"}
            runPerf = renderCharts env (fetched, jsn) tz "performance"
                         _Double
                         . filter (\s -> s /= "tas")
                         $ "ts" : map fst states
            runCont = renderCharts env (fetched, jsn) tz "contribution"
                         (_String . unpacked . _Show)
                         $ ("ts":)
                            . filter (\s -> s /="wa" && s /= "nt")
                            . map fst
                            $ states

        ((allPerfPngs, perfJSON, perfCSV)
         ,(allContPngs, contJSON, contCSV)) <- concurrently runPerf runCont

        let pngSize = foldl (\n (_,png) -> n + BSL.length (untag png))
                            0
                            (allContPngs ++ allPerfPngs)

        debugM $ "Total new PNG size: " ++ show pngSize

        let !newState = current
                & contributionGraphs    %~ H.union (H.fromList allContPngs)
                & performanceGraphs     %~ H.union (H.fromList allPerfPngs)
                & performanceCSV        .~ Just perfCSV
                & contributionCSV       .~ Just contCSV
                & performanceGraphJSON  .~ perfJSON
                & contributionGraphJSON .~ contJSON
                & latestETag            .~ (metag <|> current ^. latestETag)

        writeIORef ref $ newState
        -- performMajorGC
        return True

renderCharts :: ChartEnv -> (UTCTime,Value) -> TimeZone -> Text
             -> Prism' Value Double
             -> [Text]
             -> IO ([(Text,PngBS)], Value, Text -> CsvBS)
renderCharts env (fetched,jsn) tz title lns cols = do
    let
        vals :: [Value]
        vals = jsn ^.. key title . values

        allStates :: [(Text,[Maybe (UTCTime,Double)])]
        allStates = map (\(name,_) -> (name, getTS lns name vals)) states

        allCorrected :: [(Text,[(LocalTime,Double)])]
        allCorrected = [(name,[(utcToLocalTime tz utct,d)
                              | Just (utct,d) <- series])
                       | (name,series) <- allStates]

        allFiltered = filter (hasTwo . snd) allCorrected

        titleStr :: String
        titleStr = "All states " ++ T.unpack title

        allTitle :: Text
        allTitle = T.pack $ formatTime defaultTimeLocale (titleStr ++ " (%%) %F %X %Z") fetched

        hasTwo :: [a] -> Bool
        hasTwo (_:_:_) = True
        hasTwo _       = False

        -- allChart :: Renderable ()
        allChart = wsChart allFiltered $ do
                        layout_title L..= (T.unpack allTitle)
                        layout_y_axis . laxis_title L..= "(%)"
                        layout_x_axis . laxis_title L..= timeZoneName tz
    !allpngs <- if null allFiltered
        then return Nothing
        else do
            debugM $ "Rendering " ++ titleStr ++ " PNGs"
            allpngs' <- {-# SCC "renderCharts.renderImage(all)" #-} liftIO $ renderImage env 500 300 allChart
            return $! Just $! ("all",) $! renderToPng allpngs'


    -- spngs <- flip mapConcurrently states $ \(sname,_) -> do
    spngs <- flip mapM states $ \(sname,_) -> do
            let fullTitle = T.toUpper sname <> " " <> title <> " (%)"

                plotVals = [(utcToLocalTime tz utct,d)
                           | Just (utct,d) <- getTS lns sname vals]

                chart = wsChart [(sname,plotVals)] $ do
                            layout_title L..= (T.unpack fullTitle)
                            layout_y_axis . laxis_title L..= "(%)"
                            layout_x_axis . laxis_title L..= timeZoneName tz
            if hasTwo plotVals
                then do
                    spng' <- {-# SCC "renderCharts.renderImage(state)" #-} renderImage env 500 300 chart
                    let !spng = renderToPng spng'
                    return $ Just (sname, spng)
                else
                    return Nothing
    debugM $ "Done rendering " ++ titleStr ++ " PNGs"

    return $ (catMaybes (allpngs:spngs), renderJson vals cols, renderCSVs title allStates)


-- Produce json values for c3.js chart
renderJson :: [Value] -> [Text] -> Value
renderJson vals cols = A.toJSON $
    ((Just "ts") : [val ^? key "ts" | val <- vals]) :
    [Just (A.toJSON col) : [ val ^? key col | val <- vals] | col <- filter (/= "ts") cols]

renderCSVs :: Text -> [(Text,[Maybe (UTCTime,Double)])] -> (Text -> CsvBS)
renderCSVs title allStates =
    let tTitle = T.toTitle title
        tTitlePct = T.concat [tTitle, " (%)"]
        imageTitle = T.concat [tTitle, " over time"]

        csvHeader :: Csv.Header
        csvHeader = V.fromList [encodeUtf8 tTitlePct,"State", "State name","Received at",encodeUtf8 imageTitle] :: V.Vector S.ByteString

        currentValues :: [(Text,Maybe (UTCTime, Double))]
        currentValues = map (second maximum) allStates

        namedRecords hst =
            map (\(state, mtv)
                -> H.fromList [
                        ("State", toField $ lookup state states)
                        ,("State name", toField (T.toUpper state))
                        ,("Received at", toField $ maybe "-" (formatTime defaultTimeLocale "%F %X") (fst <$> mtv))
                        ,(encodeUtf8 tTitlePct, toField $ maybe 0.0 id (snd <$> mtv))
                        ,(encodeUtf8 imageTitle, toField $
                            T.concat ["<img src='http://",hst,"/apvi/v3/",title,"/png/",state,"'/>"])
                    ]
                )
                currentValues
        csvf hst = Tagged $ encodeByNameWith defaultEncodeOptions csvHeader (namedRecords hst)
    in csvf


getTS :: Prism' Value a -> Text -> [Value] -> [Maybe (UTCTime, a)]
getTS f state objs =
#if MIN_VERSION_time(1,5,0)
    let timeParser = parseTimeM True defaultTimeLocale "%FT%H:%M:%SZ"
#else
    let timeParser = parseTime defaultTimeLocale "%FT%H:%M:%SZ"
#endif
        timeLens   = key "ts" . _String . unpacked . to timeParser . _Just
        stateLens  = key state . f
    in sortBy (comparing (fmap fst))
       . Prelude.map (\v -> (\t s -> seq t $ seq s (t,s)) <$> v ^? timeLens <*> v ^? stateLens)
       $ objs
